snapshots/snapshots/src/main/kotlin/com/emergetools/snapshots/compose/ComposeSnapshotter.kt

```diff
@@ -170,87 +178,88 @@ private fun snapshot(
-        View.MeasureSpec.makeMeasureSpec(heightPixels, View.MeasureSpec.EXACTLY),
+  val deviceSpec = configToDeviceSpec(previewConfig)
+
+  // Use exact measurements when we have them
+  val scalingFactor = deviceSpec?.scalingFactor ?: view.resources.displayMetrics.density
+
+  val widthMeasureSpec = when {
+    previewConfig.widthDp != null -> {
+      View.MeasureSpec.makeMeasureSpec(
+        dpToPx(previewConfig.widthDp!!, scalingFactor),
+        View.MeasureSpec.EXACTLY
       )
-      return IntSize(view.measuredWidth, view.measuredHeight)
-    } else {
-      Log.e(
-        EmergeComposeSnapshotReflectiveParameterizedInvoker.TAG,
-        "Device not found for preview annotation: ${previewConfig.device}"
+    }
+
+    deviceSpec?.widthPixels != null && deviceSpec.widthPixels != -1 ->
+      View.MeasureSpec.makeMeasureSpec(deviceSpec.widthPixels, View.MeasureSpec.EXACTLY)
+
+    else ->
+      View.MeasureSpec.makeMeasureSpec(view.width, View.MeasureSpec.AT_MOST)
+  }
+
+  val heightMeasureSpec = when {
+    previewConfig.heightDp != null -> {
+      View.MeasureSpec.makeMeasureSpec(
+        dpToPx(previewConfig.heightDp!!, scalingFactor),
+        View.MeasureSpec.EXACTLY
       )
     }
+
+    deviceSpec?.heightPixels != null && deviceSpec.heightPixels != -1 ->
+      View.MeasureSpec.makeMeasureSpec(deviceSpec.heightPixels, View.MeasureSpec.EXACTLY)
+
+    else ->
+      View.MeasureSpec.makeMeasureSpec(view.height, View.MeasureSpec.AT_MOST)
   }
 
-  // Default to 0 if not set which will allow parent to impose constraints on child when
-  // AT_MOST set.
-  val emulatorDensity = view.resources.displayMetrics.density
-  val heightPx = (previewConfig.heightDp ?: 0) * emulatorDensity
-  val widthPx = (previewConfig.widthDp ?: 0) * emulatorDensity
-
-  // If width or height is set in preview annotation, measure with unspecified to allow
-  // stretching past bounds of parent.
-  // Otherwise, use AT_MOST to allow parent to impose constraints on child.
-  val widthMeasureSpec =
-    previewConfig.widthDp?.let { View.MeasureSpec.UNSPECIFIED } ?: View.MeasureSpec.AT_MOST
-  val heightMeasureSpec =
-    previewConfig.heightDp?.let { View.MeasureSpec.UNSPECIFIED } ?: View.MeasureSpec.AT_MOST
-
-  view.measure(
-    View.MeasureSpec.makeMeasureSpec(max(view.width, widthPx.toInt()), widthMeasureSpec),
-    View.MeasureSpec.makeMeasureSpec(max(view.height, heightPx.toInt()), heightMeasureSpec),
-  )
+  view.measure(widthMeasureSpec, heightMeasureSpec)
   return IntSize(view.measuredWidth, view.measuredHeight)
 }
 
+private fun updateActivityBounds(activity: Activity, deviceSpec: DeviceSpec) {
+  // Apply the device spec dimensions to the activity window
+  val width = deviceSpec.widthPixels
+  val height = deviceSpec.heightPixels
+
+  if (width > 0 && height > 0) {
+    activity.window.setLayout(width, height)
+  }
+}
+
+private fun dpToPx(dp: Int, scalingFactor: Float): Int {
+  return (dp * scalingFactor).toInt()
```

Comment:
I'm guessing this just needs to be explicitly rounded down.

Reply:
I updated to use `toInt()` in the pixel conversion later. Ended up with the proper dimensions going with that approach